From 09f4518bba3898492967a914c0c5928ef957f7ad Mon Sep 17 00:00:00 2001
From: Hua Huang <hua_huang@xyratex.com>
Date: Wed, 27 Mar 2013 18:50:53 +0800
Subject: [PATCH 04/18] error handle for cm_layout_fetch. assert layout is
 valid in layout_instance_build(). fault injection helper for UT.

---
 sns/cm/iter.c  | 28 ++++++++++++++++------------
 sns/cm/ut/cm.c | 14 ++++++++++++++
 2 files changed, 30 insertions(+), 12 deletions(-)

diff --git a/sns/cm/iter.c b/sns/cm/iter.c
index 1b236df..0ea27c6 100644
--- a/sns/cm/iter.c
+++ b/sns/cm/iter.c
@@ -334,7 +334,7 @@ static int cm_layout_fetch(struct m0_sns_cm_iter *it)
 		}
 	}
 
-	return 0;
+	return rc;
 }
 
 /*
@@ -529,21 +529,20 @@ static int layout_instance_build(struct m0_sns_cm_iter *it)
 	struct m0_layout_instance       *li;
 	struct m0_fid                   *fid;
 	int                              rc = -ENODATA;
+	M0_PRE(it != NULL);
 
 	sfc = &it->si_pl;
 	fid = &sfc->sfc_gob_fid;
 	pl  = sfc->sfc_pdlayout;
-        /* Destroy previous pdclust instance */
-	if (sfc->sfc_pi != NULL)
-		m0_layout_instance_fini(&sfc->sfc_pi->pi_base);
-	if (pl != NULL) {
-		rc = m0_layout_instance_build(&pl->pl_base.sl_base, fid, &li);
-		if (rc == 0) {
-			sfc->sfc_pi = m0_layout_instance_to_pdi(li);
-			sfc->sfc_sa.sa_group = 0;
-			sfc->sfc_sa.sa_unit = 0;
-			iter_phase_set(it, ITPH_GROUP_NEXT);
-		}
+	M0_ASSERT(pl != NULL);
+	M0_ASSERT(sfc->sfc_pi == NULL);
+
+	rc = m0_layout_instance_build(&pl->pl_base.sl_base, fid, &li);
+	if (rc == 0) {
+		sfc->sfc_pi = m0_layout_instance_to_pdi(li);
+		sfc->sfc_sa.sa_group = 0;
+		sfc->sfc_sa.sa_unit = 0;
+		iter_phase_set(it, ITPH_GROUP_NEXT);
 	}
 	return rc;
 }
@@ -600,6 +599,11 @@ static int iter_fid_next(struct m0_sns_cm_iter *it)
 		}
 
 		rc = cm_layout_fetch(it);
+		if (rc < 0 && M0_FI_ENABLED("layout_fetch_error_as_done"))
+			return -ENODATA;
+		if (rc < 0)
+			return rc;
+
 		if (rc == IT_WAIT) {
 			iter_phase_set(it, ITPH_FID_NEXT_WAIT);
 			return rc;
diff --git a/sns/cm/ut/cm.c b/sns/cm/ut/cm.c
index 3b5d9bd..7df7780 100644
--- a/sns/cm/ut/cm.c
+++ b/sns/cm/ut/cm.c
@@ -302,7 +302,9 @@ static void iter_repair_single_file(void)
 	uint64_t  fsizes[] = {ITER_64K};
 
 	iter_setup(3, 1, 5, 4096);
+	m0_fi_enable("iter_fid_next", "layout_fetch_error_as_done");
 	rc = iter_run(5, ARRAY_SIZE(fsizes), fsizes, 3, SNS_REPAIR);
+	m0_fi_disable("iter_fid_next", "layout_fetch_error_as_done");
 	M0_UT_ASSERT(rc == -ENODATA);
 	iter_stop(ARRAY_SIZE(fsizes), 5);
 }
@@ -313,7 +315,9 @@ static void iter_repair_multi_file(void)
 	uint64_t  fsizes[] = {ITER_64K, ITER_1M};
 
 	iter_setup(8, 1, 10, 4096);
+	m0_fi_enable("iter_fid_next", "layout_fetch_error_as_done");
 	rc = iter_run(10, ARRAY_SIZE(fsizes), fsizes, 4, SNS_REPAIR);
+	m0_fi_disable("iter_fid_next", "layout_fetch_error_as_done");
 	M0_UT_ASSERT(rc == -ENODATA);
 	iter_stop(ARRAY_SIZE(fsizes), 10);
 }
@@ -324,7 +328,9 @@ static void iter_repair_large_file_with_large_unit_size(void)
 	uint64_t  fsizes[] = {ITER_1G};
 
 	iter_setup(8, 1, 10, 10485760);
+	m0_fi_enable("iter_fid_next", "layout_fetch_error_as_done");
 	rc = iter_run(10, ARRAY_SIZE(fsizes), fsizes, 4, SNS_REPAIR);
+	m0_fi_disable("iter_fid_next", "layout_fetch_error_as_done");
 	M0_UT_ASSERT(rc == -ENODATA);
 	iter_stop(ARRAY_SIZE(fsizes), 10);
 }
@@ -335,7 +341,9 @@ static void iter_rebalance_single_file(void)
 	uint64_t  fsizes[] = {ITER_64K};
 
 	iter_setup(3, 1, 5, 4096);
+	m0_fi_enable("iter_fid_next", "layout_fetch_error_as_done");
 	rc = iter_run(5, ARRAY_SIZE(fsizes), fsizes, 3, SNS_REBALANCE);
+	m0_fi_disable("iter_fid_next", "layout_fetch_error_as_done");
 	M0_UT_ASSERT(rc == -ENODATA);
 	iter_stop(ARRAY_SIZE(fsizes), 5);
 }
@@ -346,7 +354,9 @@ static void iter_rebalance_multi_file(void)
 	uint64_t  fsizes[] = {ITER_64K, ITER_1M};
 
 	iter_setup(8, 1, 10, 4096);
+	m0_fi_enable("iter_fid_next", "layout_fetch_error_as_done");
 	rc = iter_run(10, ARRAY_SIZE(fsizes), fsizes, 4, SNS_REBALANCE);
+	m0_fi_disable("iter_fid_next", "layout_fetch_error_as_done");
 	M0_UT_ASSERT(rc == -ENODATA);
 	iter_stop(ARRAY_SIZE(fsizes), 10);
 }
@@ -356,7 +366,9 @@ static void iter_rebalance_large_file_with_large_unit_size(void)
 	uint64_t  fsizes[] = {ITER_1G};
 
 	iter_setup(8, 1, 10, 10485760);
+	m0_fi_enable("iter_fid_next", "layout_fetch_error_as_done");
 	rc = iter_run(10, ARRAY_SIZE(fsizes), fsizes, 4, SNS_REBALANCE);
+	m0_fi_disable("iter_fid_next", "layout_fetch_error_as_done");
 	M0_UT_ASSERT(rc == -ENODATA);
 	iter_stop(ARRAY_SIZE(fsizes), 10);
 }
@@ -367,7 +379,9 @@ static void iter_invalid_nr_cobs(void)
 	uint64_t fsizes[] = {ITER_64K};
 
 	iter_setup(3, 1, 5, 4096);
+	m0_fi_enable("iter_fid_next", "layout_fetch_error_as_done");
 	rc = iter_run(3, ARRAY_SIZE(fsizes), fsizes, 2, SNS_REPAIR);
+	m0_fi_disable("iter_fid_next", "layout_fetch_error_as_done");
 	M0_UT_ASSERT(rc == -ENODATA);
 	iter_stop(ARRAY_SIZE(fsizes), 3);
 }
-- 
1.8.3.2

